### PR TITLE
search: fix FloydWarshall

### PR DESCRIPTION
**Pull request by kortschak**
This is backported from api-explore.

An AllBetween method will be added in a later PR.

@gonum/developers Please take a look.

### CODE REVIEW

**Comment by vladimir-ch** [comment]
Argh, I wrote my comments in the commit view. Sorry.

**Comment by kortschak** [comment]
Want to replay them? This commit will go away.

**Comment by kortschak** [comment]
PTAL

File: search/floydwarshall.go [hunk 1]
```diff
@@ -5,226 +5,168 @@
...
-		}
+// ShortestPaths is a shortest path tree created by the FloydWarshall function.
+type ShortestPaths struct {
+	nodes   []graph.Node
+	indexOf map[int]int
 
-		return paths, dist[start.ID()+goal.ID()*numNodes], nil
-	}
+	dist *mat64.Dense
+	next [][]int
```
Comment (vladimir-ch) [inline search/floydwarshall.go]: I would add a comment explaining what next[i] contains. It is obvious to me now, but the understanding was definitely not immedate. The algorithm then follows easily.

File: search/floydwarshall_test.go [hunk 2]
```diff
@@ -1,182 +1,266 @@
-// Copyright ©2014 The gonum Authors. All rights reserved.
+// Copyright ©2014 The gonum Authors. AllBetween rights reserved.
```
Comment (vladimir-ch) [inline search/floydwarshall_test.go]: AllBetween rights :-)

File: search/floydwarshall.go (hunk 1 above)
Comment (kortschak) [inline search/floydwarshall.go]: Done. Also added full comments to other fields here.

File: search/floydwarshall.go [hunk 3]
```diff
@@ -5,226 +5,168 @@
...
-	}
-
-	for _, node := range nodes {
-		for _, succ := range successors(node) {
-			dist[node.ID()+succ.ID()*numNodes] = cost(edgeTo(node, succ))
-		}
+// FloydWarshall returns a shortest path tree for the graph g or false indicating
+// that a negative cycle exists in the graph. If weight is nil and the graph does not
+// implement graph.Coster, UniformCost is used.
+func FloydWarshall(g graph.Graph, weight graph.CostFunc) (paths ShortestPaths, ok bool) {
```
Comment (vladimir-ch) [inline search/floydwarshall.go]: Perhaps a stupid question: how common is the situation of having a DirectedGraph but wanting to use FloydWarshall as if it were just an undirected Graph? It seems that one would have to create a kind of wrapper in such situation. Alternatively, there could be two functions, e.g. FloydWarshall() and FloydWarshallUndirected(), that would take the corresponding interface, wouldn't do any type switching (except temporarily on graph.Coster) and would just call a private method with nodes, weight, from, edgeTo (from line 39 on that's all that is needed).

File: search/floydwarshall_test.go (hunk 2 above)
Comment (kortschak) [inline search/floydwarshall_test.go]: s/.*/Something meaningless/g

File: search/floydwarshall.go (hunk 3 above)
Comment (kortschak) [inline search/floydwarshall.go]: I'm not sure I understand. The current situation results from the fact that there is no interface that differentiates an undirected graph from a general graph, so undirected is the default. This is not the case in the api-explore branch. I would rather not split this into two functions, the type switching is not costly and happens only once.

Comment (vladimir-ch) [inline search/floydwarshall.go]: Ah, ok, I understand. As I wrote, bad question.

Comment (kortschak) [inline search/floydwarshall.go]: Ideally, I'd like the CostFunc to go away too. It's the next thing on my shopping list at home. I think it should be dealt with by the graph interfaces.

**Comment by vladimir-ch** [comment]
I agree with removing Crunch interface and I checked floydwarshall.go in detail, it LGTM. I did not check the tests thoroughly, though, I will trust you on them.

**Comment by kortschak** [comment]
Thank you for the review.

**Pull request merged**